src/core/ddsi/src/ddsi_raweth.c

```diff
@@ -391,6 +395,334 @@ static int joinleave_asm_mcgroup (ddsrt_socket_t socket, int join, const ddsi_lo
   return (rc == DDS_RETCODE_OK) ? 0 : rc;
 }
 
+#elif DDSI_USE_BSD
+
+#define DEFAULT_BUFFER_SIZE 1000000
+
+struct ddsi_vlan_tag {
+  unsigned short tag;
+  unsigned short proto; 
+};
+
+static ssize_t ddsi_raweth_conn_read (struct ddsi_tran_conn * conn, unsigned char * buf, size_t len, bool allow_spurious, ddsi_locator_t *srcloc)
+{
+  ssize_t ret  = 0;
+  dds_return_t rc = DDS_RETCODE_OK;
+  ddsi_raweth_conn_t uc = (ddsi_raweth_conn_t) conn;
+  struct bpf_hdr *bpf_hdr;
+  struct ddsi_ethernet_header *eth_hdr;
+  struct ddsi_vlan_tag *vtag = NULL;
+  uint32_t port = 0;
+  char *ptr;
+  (void) allow_spurious;
+
+  ddsrt_mutex_lock (&uc->lock);
+
+  if (uc->avail == 0)
+  {
+    if ((ret = read(uc->m_sock, uc->buffer, uc->buflen)) <= 0)
+    {
+      DDS_CERROR (&conn->m_base.gv->logconfig, "ddsi_raweth_create_conn read failed ... retcode = %"PRIdSIZE"\n", ret);
+      rc = DDS_RETCODE_ERROR;
+      goto error;
+    }
+    uc->avail = ret;
+    uc->bptr = uc->buffer;
+  }
+
+  if (uc->bptr < uc->buffer + uc->avail)
```

Comment:
For all the `bptr` manipulations/checks: these structures come straight from the OS kernel, so I suppose we can trust them. I do think it would be good to point out in a comment that this is the reason there is so little checking, to help any future reader.